azure-iot-device/azure/iot/device/iothub/aio/async_handler_manager.py

```diff
@@ -0,0 +1,151 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+""" This module contains the manager for handler methods used by the aio clients"""
+
+from azure.iot.device.common import asyncio_compat
+from azure.iot.device.common.chainable_exception import ChainableException
+from concurrent.futures import ThreadPoolExecutor
+import asyncio
+import logging
+import inspect
+
+logger = logging.getLogger(__name__)
+
+MESSAGE = "_on_message_received"
+METHOD = "_on_method_request_received"
+TWIN_DP_PATCH = "_on_twin_desired_properties_patch_received"
+# TODO: add more for "event"
+
+
+class AsyncHandlerManagerException(ChainableException):
+    pass
+
+
+class AsyncHandlerManager(object):
+    def __init__(self, inbox_manager):
+        self._inbox_manager = inbox_manager
+
+        self._handler_tasks = {
+            # Inbox handler tasks
+            MESSAGE: None,
+            METHOD: None,
+            TWIN_DP_PATCH: None,
+            # Other handler tasks
+            # TODO: add
+        }
+
+        # Inbox handlers
+        self._on_message_received = None
+        self._on_method_request_received = None
+        self._on_twin_desired_properties_patch_received = None
+
+        # Other handlers
+        # TODO: add
+
+    async def _run_inbox_handler(self, inbox, handler_name):
+        """Run infinite loop that waits for an inbox to receive an object from it, then calls
+        the handler with that object
+        """
+        # Run the handler in a threadpool, so that it cannot block other handlers (from a different task),
+        # or the main client thread. The number of worker threads forms an upper bound on how many instances
+        # of the same handler can be running simultaneously.
+        tp = ThreadPoolExecutor(max_workers=4)
+        while True:
+            retval = await inbox.get()
+            # NOTE: we MUST use getattr here using the handler name, as opposed to directly passing
+            # the handler in order for the handler to be able to be updated without cancelling
+            # the running task created for this coroutine
+            handler = getattr(self, handler_name)
+            if inspect.iscoroutinefunction(handler):
+                # Wrap the coroutine in a function so it can be run in ThreadPool
+                def coro_wrapper(coro, arg):
+                    asyncio_compat.run(coro(arg))
+
+                tp.submit(coro_wrapper, handler, retval)
```

Comment:
>tp.submit(coro_wrapper, handler, retval) [](start = 16, length = 40)

Please don't change this since this code is going to be updated, but this line will swallow any errors that get raised from inside coro_wrapper.  It's a real hassle to get exceptions that are raised from threads.  Fortunately, the threadpool makes this easier by returning a concurrent.futures.Future object (which is not an asyncio.Future object.  what a wonderful world we live in.)   [https://stackoverflow.com/questions/2829329/catch-a-threads-exception-in-the-caller-thread-in-python](https://stackoverflow.com/questions/2829329/catch-a-threads-exception-in-the-caller-thread-in-python)